pilotecmwf_pp_starter/__main__.py

```diff
@@ -1,13 +1,200 @@
 
-_LOGGER = logging.getLogger(__name__)
+# Initialize logging
+logging.basicConfig(level=logging.INFO)
 
+# Configuration
+GB = 1024 ** 3
+TRANSFER_CONFIG = TransferConfig(multipart_threshold=5 * GB)
+S3INPUT_ENDPOINT_URL = "https://object-store.os-api.cci1.ecmwf.int"
+S3OUTPUT_ENDPOINT_URL = "" # NB: doesn't exist yet
+S3INPUT_BUCKET_NAME = "flexpart-input"
+S3OUTPUT_BUCKET_NAME = "flexpart-output" # NB: doesn't exist yet
+TINCR = int(os.getenv('TINCR', 3))
+TSTART = int(os.getenv('TSTART', 0))
 
-def simulate_job():
-    _LOGGER.info("Here we are. Now going to sleep.")
-    time.sleep(5)
-    _LOGGER.info("Sleeping finished. Back to work.")
+def get_s3_client(endpoint_url, access_key, secret_key):
+    return boto3.client(
+        's3',
+        endpoint_url=endpoint_url,
+        aws_access_key_id=access_key,
+        aws_secret_access_key=secret_key,
+        use_ssl=True,
+    )
 
+s3_input = get_s3_client(S3INPUT_ENDPOINT_URL, os.getenv('S3INPUT_ACCESS_KEY'), os.getenv('S3INPUT_SECRET_KEY'))
+# NB: doesn't exist yet
+# s3_output = get_s3_client(S3OUTPUT_ENDPOINT_URL, os.getenv('S3OUTPUT_ACCESS_KEY'), os.getenv('S3OUTPUT_SECRET_KEY'))
+
+# Define constants and input fields for flexpart
+constants = ("z", "lsm", "sdor")
+input_fields = (
+    "u", "v", "etadot", "t", "q", "sp", "10u", "10v", "2t", "2d", "tcc", "sd",
+    "cp", "lsp", "ssr", "sshf", "ewss", "nsss"
+)
+
+def download_file(s3_client, bucket, key, local_path):
+    try:
+        s3_client.download_file(bucket, key, local_path, Config=TRANSFER_CONFIG)
+        logging.info(f"Downloaded file from S3: {key}")
+    except ClientError as e:
+        logging.error(f"Error downloading file {key}: {e}")
+
+def upload_file(s3_client, bucket, local_path):
+    key = os.path.basename(local_path)
+    try:
+        s3_client.upload_file(local_path, bucket, key)
+        logging.info(f"Uploaded file to S3: {key}")
+    except ClientError as e:
+        logging.error(f"Error uploading file {local_path}: {e}")
+
+def validate_dataset(ds, params, ref_time, step, prev_step):
+    if not all(param in ds.keys() for param in params):
+        raise ValueError("Not all requested parameters are present in the dataset")
+    if not all(
+        (np.array_equal(array.coords["time"].values, [0])) or
+        (prev_step == 0 and np.array_equal(array.coords["time"].values, [0, step])) or
+        (prev_step != 0 and np.array_equal(array.coords["time"].values, [0, prev_step, step]))
+        for array in ds.values()
+    ):
+        raise ValueError("Downloaded steps are incorrect")
+    if not all((array.coords["ref_time"].values == np.datetime64(ref_time, 'ns')) for array in ds.values()):
+        raise ValueError("The forecast reference time is incorrect")
+
+def process_fields(ds_out, ds_in, input_fields, constant_fields):
+    missing_fields = [field for field in ds_in.keys() if field not in ds_out.keys() and field in input_fields and field != 'etadot']
```

Comment:
This should be a set operation
```suggestion
    missing_fields = (ds_in.keys() & input_fields)  - {"etadot"} - ds_out.keys()
```
note that `input_fields` needs to be a set in this case